File: app/financials/financial_reports_screen.py
import datetime
import random

class FinancialReportsScreen:
    def __init__(self):
        # These would be aggregates from other systems (accounting, sales, expenses)
        self.financial_data = self._generate_mock_financial_data()

    def _generate_mock_financial_data(self, year=datetime.date.today().year, month=datetime.date.today().month):
        data = {"year": year, "month": month}

        # --- Income Statement (Profit & Loss) Data ---
        data["revenue"] = {
            "consultation_fees": round(random.uniform(50000, 150000), 2),
            "marketplace_commissions": round(random.uniform(30000, 100000), 2),
            "subscription_fees": round(random.uniform(10000, 50000), 2),
            "other_income": round(random.uniform(1000, 5000), 2),
        }
        data["total_revenue"] = sum(data["revenue"].values())

        data["cost_of_goods_sold"] = { # COGS, if applicable (e.g. for physical products if sold directly)
            "platform_service_costs": round(data["total_revenue"] * random.uniform(0.05, 0.10), 2), # Costs directly tied to delivering services
        }
        data["total_cogs"] = sum(data["cost_of_goods_sold"].values())
        data["gross_profit"] = data["total_revenue"] - data["total_cogs"]

        data["operating_expenses"] = {
            "salaries_and_wages": round(random.uniform(20000, 60000), 2),
            "marketing_and_advertising": round(random.uniform(5000, 20000), 2),
            "software_and_tools": round(random.uniform(2000, 10000), 2),
            "rent_and_utilities": round(random.uniform(1000, 5000), 2),
            "payment_processing_fees": round(data["total_revenue"] * random.uniform(0.01, 0.03), 2),
            "customer_support": round(random.uniform(3000, 8000), 2),
            "other_operating_expenses": round(random.uniform(1000, 4000), 2),
        }
        data["total_operating_expenses"] = sum(data["operating_expenses"].values())
        data["operating_income"] = data["gross_profit"] - data["total_operating_expenses"] # EBIT

        data["non_operating"] = { # Interest, Taxes
            "interest_expense": round(data["operating_income"] * random.uniform(0, 0.02) if data["operating_income"] > 0 else 0, 2),
            "taxes": round(data["operating_income"] * random.uniform(0.1,0.2) if data["operating_income"] > 0 else 0, 2), # Simplified tax
        }
        data["net_income"] = data["operating_income"] - sum(data["non_operating"].values())


        # --- Balance Sheet Data (Simplified Summary) ---
        # Assets
        current_assets_cash = round(random.uniform(100000, 500000), 2)
        current_assets_receivables = round(data["total_revenue"] * random.uniform(0.1, 0.2), 2) # e.g. outstanding marketplace fees
        data["assets"] = {
            "current_assets": {
                "cash_and_equivalents": current_assets_cash,
                "accounts_receivable": current_assets_receivables,
                "prepaid_expenses": round(random.uniform(5000, 20000), 2),
            },
            "non_current_assets": {
                "property_plant_equipment_net": round(random.uniform(20000, 100000), 2),
                "intangible_assets_net": round(random.uniform(10000, 50000), 2), # e.g. software development costs capitalized
            }
        }
        data["total_current_assets"] = sum(data["assets"]["current_assets"].values())
        data["total_non_current_assets"] = sum(data["assets"]["non_current_assets"].values())
        data["total_assets"] = data["total_current_assets"] + data["total_non_current_assets"]

        # Liabilities
        current_liabilities_payable = round(data["total_operating_expenses"] * random.uniform(0.1, 0.15),2) # e.g. unpaid bills
        current_liabilities_deferred_revenue = round(data["revenue"]["subscription_fees"] * random.uniform(0.2,0.4),2) # unearned subscription revenue
        data["liabilities"] = {
            "current_liabilities": {
                "accounts_payable": current_liabilities_payable,
                "deferred_revenue": current_liabilities_deferred_revenue, # Subscriptions paid upfront
                "accrued_expenses": round(random.uniform(3000, 15000), 2),
            },
            "non_current_liabilities": {
                "long_term_debt": round(random.uniform(0, 50000), 2),
            }
        }
        data["total_current_liabilities"] = sum(data["liabilities"]["current_liabilities"].values())
        data["total_non_current_liabilities"] = sum(data["liabilities"]["non_current_liabilities"].values())
        data["total_liabilities"] = data["total_current_liabilities"] + data["total_non_current_liabilities"]

        # Equity (Derived: Assets - Liabilities)
        # In a real system, equity would have components like common stock, retained earnings
        data["total_equity"] = data["total_assets"] - data["total_liabilities"]
        data["equity"] = {
            "retained_earnings_placeholder": data["total_equity"] - 50000, # Assuming 50k is common stock
            "common_stock_placeholder": 50000,
        }

        return data

    def generate_profit_and_loss_statement(self, period_data):
        report = f"--- Profit and Loss Statement ---\n"
        report += f"For Period Ending: {period_data['month']}/{period_data['year']}\n\n" # Simplified, could be more specific (e.g. "Month of...")

        report += "Revenue:\n"
        for item, amount in period_data["revenue"].items():
            report += f"  {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"Total Revenue: ${period_data['total_revenue']:,.2f}\n\n"

        report += "Cost of Goods Sold:\n"
        for item, amount in period_data["cost_of_goods_sold"].items():
            report += f"  {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"Total COGS: ${period_data['total_cogs']:,.2f}\n"
        report += f"Gross Profit: ${period_data['gross_profit']:,.2f}\n\n"

        report += "Operating Expenses:\n"
        for item, amount in period_data["operating_expenses"].items():
            report += f"  {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"Total Operating Expenses: ${period_data['total_operating_expenses']:,.2f}\n"
        report += f"Operating Income (EBIT): ${period_data['operating_income']:,.2f}\n\n"

        report += "Non-Operating Income/Expenses:\n"
        report += f"  Interest Expense: ${period_data['non_operating']['interest_expense']:,.2f}\n"
        report += f"Taxes: ${period_data['non_operating']['taxes']:,.2f}\n"
        report += f"Net Income: ${period_data['net_income']:,.2f}\n"

        return report

    def generate_balance_sheet_summary(self, period_data):
        report = f"--- Balance Sheet Summary ---\n"
        report += f"As of: {period_data['month']}/{period_data['year']}\n\n" # Simplified

        report += "Assets:\n"
        report += "  Current Assets:\n"
        for item, amount in period_data["assets"]["current_assets"].items():
            report += f"    {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"  Total Current Assets: ${period_data['total_current_assets']:,.2f}\n"
        report += "  Non-Current Assets:\n"
        for item, amount in period_data["assets"]["non_current_assets"].items():
            report += f"    {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"  Total Non-Current Assets: ${period_data['total_non_current_assets']:,.2f}\n"
        report += f"Total Assets: ${period_data['total_assets']:,.2f}\n\n"

        report += "Liabilities:\n"
        report += "  Current Liabilities:\n"
        for item, amount in period_data["liabilities"]["current_liabilities"].items():
            report += f"    {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"  Total Current Liabilities: ${period_data['total_current_liabilities']:,.2f}\n"
        report += "  Non-Current Liabilities:\n"
        for item, amount in period_data["liabilities"]["non_current_liabilities"].items():
            report += f"    {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"  Total Non-Current Liabilities: ${period_data['total_non_current_liabilities']:,.2f}\n"
        report += f"Total Liabilities: ${period_data['total_liabilities']:,.2f}\n\n"

        report += "Equity:\n"
        # Simplified display for equity components
        for item, amount in period_data["equity"].items():
             report += f"  {item.replace('_', ' ').title()}: ${amount:,.2f}\n"
        report += f"Total Equity: ${period_data['total_equity']:,.2f}\n\n"

        report += f"Total Liabilities and Equity: ${period_data['total_liabilities'] + period_data['total_equity']:,.2f}\n"
        # Sanity check: Total Assets should equal Total Liabilities + Total Equity
        if abs((period_data['total_liabilities'] + period_data['total_equity']) - period_data['total_assets']) > 0.01: # Check for floating point issues
            report += "Error: Balance sheet does not balance!\n"

        return report

    def display_screen(self):
        print("---- Financial Reports Screen ----")

        # For current period (e.g., current month/year based on mock data generation)
        current_period_data = self.financial_data

        print("\nGenerating Profit and Loss Statement for current period...")
        pnl_report = self.generate_profit_and_loss_statement(current_period_data)
        print(pnl_report)

        print("\nGenerating Balance Sheet Summary for current period...")
        bs_summary_report = self.generate_balance_sheet_summary(current_period_data)
        print(bs_summary_report)

        # Example: Generating for a previous period (would require historical data storage in real app)
        # print("\nGenerating report for a previous period (mocked by regenerating data)...")
        # previous_period_data = self._generate_mock_financial_data(year=self.financial_data["year"], month=self.financial_data["month"]-1 if self.financial_data["month"] > 1 else 12)
        # pnl_report_prev = self.generate_profit_and_loss_statement(previous_period_data)
        # print(pnl_report_prev)

        print("---------------------------------")

if __name__ == '__main__':
    screen = FinancialReportsScreen()
    screen.display_screen()

File: app/src/main/java/com/example/rooster/financials/reports/FinancialReportsScreen.kt
package com.example.rooster.financials.reports

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.text.DecimalFormat
import java.util.Calendar
import kotlin.random.Random

// --- Data Classes ---
// Using Map<String, Double> for simplicity, similar to Python. Could be more structured.
data class FinancialReportData(
    val year: Int,
    val month: Int,
    val revenue: Map<String, Double>,
    val totalRevenue: Double,
    val costOfGoodsSold: Map<String, Double>,
    val totalCogs: Double,
    val grossProfit: Double,
    val operatingExpenses: Map<String, Double>,
    val totalOperatingExpenses: Double,
    val operatingIncome: Double, // EBIT
    val nonOperating: Map<String, Double>, // Interest, Taxes
    val netIncome: Double,
    // Balance Sheet parts
    val currentAssets: Map<String, Double>,
    val totalCurrentAssets: Double,
    val nonCurrentAssets: Map<String, Double>,
    val totalNonCurrentAssets: Double,
    val totalAssets: Double,
    val currentLiabilities: Map<String, Double>,
    val totalCurrentLiabilities: Double,
    val nonCurrentLiabilities: Map<String, Double>,
    val totalNonCurrentLiabilities: Double,
    val totalLiabilities: Double,
    val equity: Map<String, Double>, // Placeholder for actual equity components
    val totalEquity: Double
)

// --- ViewModel ---
class FinancialReportsViewModel : ViewModel() {
    private val _financialReportData = MutableStateFlow<FinancialReportData?>(null)
    val financialReportData: StateFlow<FinancialReportData?> = _financialReportData

    init {
        loadMockFinancialData()
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadMockFinancialData() {
        val today = Calendar.getInstance()
        val year = today.get(Calendar.YEAR)
        val month = today.get(Calendar.MONTH) + 1 // Calendar.MONTH is 0-indexed

        val revenueMap = mapOf(
            "consultation_fees" to Random.nextDouble(50000.0, 150000.0),
            "marketplace_commissions" to Random.nextDouble(30000.0, 100000.0),
            "subscription_fees" to Random.nextDouble(10000.0, 50000.0),
            "other_income" to Random.nextDouble(1000.0, 5000.0)
        )
        val totalRevenue = revenueMap.values.sum()

        val cogsMap = mapOf(
            "platform_service_costs" to totalRevenue * Random.nextDouble(0.05, 0.10)
        )
        val totalCogs = cogsMap.values.sum()
        val grossProfit = totalRevenue - totalCogs

        val opExpensesMap = mapOf(
            "salaries_and_wages" to Random.nextDouble(20000.0, 60000.0),
            "marketing_and_advertising" to Random.nextDouble(5000.0, 20000.0),
            "software_and_tools" to Random.nextDouble(2000.0, 10000.0),
            "rent_and_utilities" to Random.nextDouble(1000.0, 5000.0),
            "payment_processing_fees" to totalRevenue * Random.nextDouble(0.01, 0.03),
            "customer_support" to Random.nextDouble(3000.0, 8000.0),
            "other_operating_expenses" to Random.nextDouble(1000.0, 4000.0)
        )
        val totalOpExpenses = opExpensesMap.values.sum()
        val operatingIncome = grossProfit - totalOpExpenses

        val nonOpMap = mapOf(
            "interest_expense" to (if (operatingIncome > 0) operatingIncome * Random.nextDouble(0.0, 0.02) else 0.0),
            "taxes" to (if (operatingIncome > 0) operatingIncome * Random.nextDouble(0.1, 0.2) else 0.0)
        )
        val netIncome = operatingIncome - nonOpMap.values.sum()

        // Balance Sheet
        val currentAssetsMap = mapOf(
            "cash_and_equivalents" to Random.nextDouble(100000.0, 500000.0),
            "accounts_receivable" to totalRevenue * Random.nextDouble(0.1, 0.2),
            "prepaid_expenses" to Random.nextDouble(5000.0, 20000.0)
        )
        val totalCurrentAssets = currentAssetsMap.values.sum()
        val nonCurrentAssetsMap = mapOf(
            "property_plant_equipment_net" to Random.nextDouble(20000.0, 100000.0),
            "intangible_assets_net" to Random.nextDouble(10000.0, 50000.0)
        )
        val totalNonCurrentAssets = nonCurrentAssetsMap.values.sum()
        val totalAssets = totalCurrentAssets + totalNonCurrentAssets

        val currentLiabilitiesMap = mapOf(
            "accounts_payable" to totalOpExpenses * Random.nextDouble(0.1, 0.15),
            "deferred_revenue" to revenueMap.getOrDefault("subscription_fees", 0.0) * Random.nextDouble(0.2, 0.4),
            "accrued_expenses" to Random.nextDouble(3000.0, 15000.0)
        )
        val totalCurrentLiabilities = currentLiabilitiesMap.values.sum()
        val nonCurrentLiabilitiesMap = mapOf(
            "long_term_debt" to Random.nextDouble(0.0, 50000.0)
        )
        val totalNonCurrentLiabilities = nonCurrentLiabilitiesMap.values.sum()
        val totalLiabilities = totalCurrentLiabilities + totalNonCurrentLiabilities

        val totalEquity = totalAssets - totalLiabilities
        val equityMap = mapOf(
            "retained_earnings_placeholder" to totalEquity - 50000.0, // Assuming 50k common stock
            "common_stock_placeholder" to 50000.0
        )


        _financialReportData.value = FinancialReportData(
            year, month, revenueMap, totalRevenue, cogsMap, totalCogs, grossProfit,
            opExpensesMap, totalOpExpenses, operatingIncome, nonOpMap, netIncome,
            currentAssetsMap, totalCurrentAssets, nonCurrentAssetsMap, totalNonCurrentAssets, totalAssets,
            currentLiabilitiesMap, totalCurrentLiabilities, nonCurrentLiabilitiesMap, totalNonCurrentLiabilities, totalLiabilities,
            equityMap, totalEquity
        )
    }
}

// --- Composable Screen ---
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FinancialReportsScreen(viewModel: FinancialReportsViewModel = androidx.lifecycle.viewmodel.compose.viewModel()) {
    val reportData by viewModel.financialReportData.collectAsState()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Financial Reports") },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFF42A5F5)) // Blue
            )
        }
    ) { paddingValues ->
        LazyColumn(modifier = Modifier.padding(paddingValues).padding(16.dp)) {
            reportData?.let { data ->
                item { ProfitAndLossStatement(data) }
                item { Spacer(modifier = Modifier.height(24.dp)) }
                item { BalanceSheetSummary(data) }
            } ?: item {
                Box(modifier = Modifier.fillMaxSize(), contentAlignment = androidx.compose.ui.Alignment.Center) {
                    CircularProgressIndicator()
                }
            }
        }
    }
}

val decimalFormat = DecimalFormat("$ #,##0.00;($ #,##0.00)") // Positive;Negative format

@Composable
fun ReportRow(label: String, amount: Double?, isTotal: Boolean = false, indentLevel: Int = 0) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(start = (indentLevel * 16).dp, vertical = 2.dp),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = label.replace("_", " ").replaceFirstChar { it.titlecase() },
            fontWeight = if (isTotal) FontWeight.Bold else FontWeight.Normal,
            fontSize = if (isTotal) 15.sp else 14.sp
        )
        Text(
            text = amount?.let { decimalFormat.format(it) } ?: "-",
            fontWeight = if (isTotal) FontWeight.Bold else FontWeight.Normal,
            fontSize = if (isTotal) 15.sp else 14.sp
        )
    }
}

@Composable
fun ReportSectionHeader(title: String) {
    Text(
        text = title,
        style = MaterialTheme.typography.titleMedium,
        modifier = Modifier.padding(top = 12.dp, bottom = 6.dp)
    )
}


@Composable
fun ProfitAndLossStatement(data: FinancialReportData) {
    Card(elevation = CardDefaults.cardElevation(2.dp), modifier = Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp)) {
            Text("Profit and Loss Statement", style = MaterialTheme.typography.headlineSmall)
            Text("For Period Ending: ${data.month}/${data.year}", style = MaterialTheme.typography.bodySmall)
            Divider(modifier = Modifier.padding(vertical = 8.dp))

            ReportSectionHeader("Revenue")
            data.revenue.forEach { (key, value) -> ReportRow(key, value, indentLevel = 1) }
            ReportRow("Total Revenue", data.totalRevenue, isTotal = true)

            ReportSectionHeader("Cost of Goods Sold")
            data.costOfGoodsSold.forEach { (key, value) -> ReportRow(key, value, indentLevel = 1) }
            ReportRow("Total COGS", data.totalCogs, isTotal = true)
            ReportRow("Gross Profit", data.grossProfit, isTotal = true, indentLevel = -1) // Negative indent to align with major totals

            ReportSectionHeader("Operating Expenses")
            data.operatingExpenses.forEach { (key, value) -> ReportRow(key, value, indentLevel = 1) }
            ReportRow("Total Operating Expenses", data.totalOperatingExpenses, isTotal = true)
            ReportRow("Operating Income (EBIT)", data.operatingIncome, isTotal = true, indentLevel = -1)

            ReportSectionHeader("Non-Operating Income/Expenses")
            data.nonOperating.forEach { (key, value) -> ReportRow(key, value, indentLevel = 1) }
            ReportRow("Net Income", data.netIncome, isTotal = true, indentLevel = -1)
        }
    }
}

@Composable
fun BalanceSheetSummary(data: FinancialReportData) {
    Card(elevation = CardDefaults.cardElevation(2.dp), modifier = Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp)) {
            Text("Balance Sheet Summary", style = MaterialTheme.typography.headlineSmall)
            Text("As of: ${data.month}/${data.year}", style = MaterialTheme.typography.bodySmall)
            Divider(modifier = Modifier.padding(vertical = 8.dp))

            ReportSectionHeader("Assets")
            ReportSectionHeader("Current Assets")
            data.currentAssets.forEach { (key, value) -> ReportRow(key, value, indentLevel = 2) }
            ReportRow("Total Current Assets", data.totalCurrentAssets, isTotal = true, indentLevel = 1)

            ReportSectionHeader("Non-Current Assets")
            data.nonCurrentAssets.forEach { (key, value) -> ReportRow(key, value, indentLevel = 2) }
            ReportRow("Total Non-Current Assets", data.totalNonCurrentAssets, isTotal = true, indentLevel = 1)
            ReportRow("Total Assets", data.totalAssets, isTotal = true)

            ReportSectionHeader("Liabilities")
            ReportSectionHeader("Current Liabilities")
            data.currentLiabilities.forEach { (key, value) -> ReportRow(key, value, indentLevel = 2) }
            ReportRow("Total Current Liabilities", data.totalCurrentLiabilities, isTotal = true, indentLevel = 1)

            ReportSectionHeader("Non-Current Liabilities")
            data.nonCurrentLiabilities.forEach { (key, value) -> ReportRow(key, value, indentLevel = 2) }
            ReportRow("Total Non-Current Liabilities", data.totalNonCurrentLiabilities, isTotal = true, indentLevel = 1)
            ReportRow("Total Liabilities", data.totalLiabilities, isTotal = true)

            ReportSectionHeader("Equity")
            data.equity.forEach { (key, value) -> ReportRow(key, value, indentLevel = 1) }
            ReportRow("Total Equity", data.totalEquity, isTotal = true)
            Divider(modifier = Modifier.padding(vertical = 4.dp))
            ReportRow("Total Liabilities and Equity", data.totalLiabilities + data.totalEquity, isTotal = true)

            // Sanity Check (visual cue, actual logic would be in VM or tests)
            if (kotlin.math.abs((data.totalLiabilities + data.totalEquity) - data.totalAssets) > 0.01) {
                Text("Error: Balance sheet does not balance!", color = Color.Red, fontWeight = FontWeight.Bold)
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
fun PreviewFinancialReportsScreen() {
    MaterialTheme {
        FinancialReportsScreen(viewModel = FinancialReportsViewModel())
    }
}
